compiler/rustc_trait_selection/src/traits/chalk_fulfill.rs

```diff
@@ -4,12 +4,16 @@ use crate::infer::canonical::OriginalQueryValues;
 use crate::infer::InferCtxt;
 use crate::traits::query::NoSolution;
 use crate::traits::{
-    ChalkEnvironmentAndGoal, ChalkEnvironmentClause, FulfillmentError, FulfillmentErrorCode,
-    ObligationCause, PredicateObligation, SelectionError, TraitEngine,
+    ChalkGoal as Goal,
```

Comment:
This going "forth and back" between `Goal` and `ChalkGoal` is a bit awkward. It's called `Goal` if the module is already chalk specific, such as `chalk_fulfill`, and `ChalkGoal` otherwise.

Reply:
Might be easier to just stick to one name for now. We can always rename it later.

Reply:
Which one would you prefer? I thought it was [a bad habit to repeat part of the module name](https://rust-lang.github.io/rust-clippy/master/index.html#module_name_repetitions), so perhaps `Goal`?

Reply:
(that lint was basically the reason I chose this solution in the first place; I think it is clear, but the naming goes forth and back to `ChalkGoal` because of the "route" of the imports)

Reply:
Since we're still integrating chalk, `ChalkGoal` should be fine for now to avoid confusion (even if it's redundant in some places)